File: app/src/main/java/me/myshows/android/ui/activities/LoginActivity.java
package me.myshows.android.ui.activities;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.TextPaint;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.trello.rxlifecycle.ActivityEvent;
import com.trello.rxlifecycle.components.support.RxAppCompatActivity;

import me.myshows.android.MyShowsApplication;
import me.myshows.android.R;
import me.myshows.android.api.MyShowsClient;
import me.myshows.android.api.impl.Credentials;
import rx.Single;

/**
 * @author Whiplash
 * @date 14.06.2015
 */
public class LoginActivity extends RxAppCompatActivity {

    private static final String REGISTER_URL = "http://myshows.me/";
    private static final int ANIMATION_DURATION = 500;

    private MyShowsClient client;

    private View logo;
    private ViewGroup loginLayout;

    private boolean needAnimate;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE |
                    View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN |
                    View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION);
        }

        client = MyShowsApplication.getMyShowsClient(this);

        logo = findViewById(R.id.logo);
        loginLayout = (ViewGroup) findViewById(R.id.login_layout);

        TextView newAccount = (TextView) findViewById(R.id.new_account);
        setupNewAccountTextView(newAccount);

        findViewById(R.id.sign_in_button).setOnClickListener(view -> {
            String login = ((EditText) findViewById(R.id.login)).getText().toString();
            String password = ((EditText) findViewById(R.id.password)).getText().toString();
            Credentials credentials = Credentials.make(login, password);
            processAuthenticationObserver(client.authentication(credentials));
        });

        autoLoginAttempt();
    }

    private void autoLoginAttempt() {
        if (client.hasCredentials()) {
            if (hasInternetConnection()) {
                needAnimate = true;
                setEnabled(loginLayout, false);
                processAuthenticationObserver(client.autoAuthentication());
            } else {
                changeActivity();
            }
        } else {
            logo.setTranslationY(0);
            loginLayout.setAlpha(1);
        }
    }

    private void setupNewAccountTextView(TextView view) {
        String register = getString(R.string.register);
        String newAccount = getString(R.string.new_account, register);
        int start = newAccount.indexOf(register);
        int end = start + register.length();
        SpannableString ss = new SpannableString(newAccount);
        ClickableSpan clickableSpan = new ClickableSpan() {
            @Override
            public void onClick(View widget) {
                Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(REGISTER_URL));
                startActivity(intent);
            }

            @Override
            public void updateDrawState(TextPaint ds) {
                ds.setColor(ContextCompat.getColor(LoginActivity.this, R.color.red_80_opacity));
                ds.setUnderlineText(true);
            }
        };
        ss.setSpan(clickableSpan, start, end, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        view.setText(ss, TextView.BufferType.SPANNABLE);
        view.setMovementMethod(LinkMovementMethod.getInstance());
    }

    private void setEnabled(ViewGroup parent, boolean enabled) {
        parent.setEnabled(enabled);
        for (int i = 0; i < parent.getChildCount(); i++) {
            View child = parent.getChildAt(i);
            if (child instanceof ViewGroup) {
                setEnabled((ViewGroup) child, enabled);
            } else {
                child.setEnabled(enabled);
            }
        }
    }

    private void animate() {
        ObjectAnimator logoAnimator = ObjectAnimator.ofFloat(logo, View.TRANSLATION_Y, 0);
        ObjectAnimator loginLayoutAnimator = ObjectAnimator.ofFloat(loginLayout, View.ALPHA, 1);
        AnimatorSet animation = new AnimatorSet();
        animation.playTogether(logoAnimator, loginLayoutAnimator);
        animation.setDuration(ANIMATION_DURATION);
        animation.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                setEnabled(loginLayout, true);
                needAnimate = false;
            }
        });
        animation.start();
    }

    private boolean hasInternetConnection() {
        ConnectivityManager connectivityManager =
                (ConnectivityManager) getApplicationContext().getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo netInfo = connectivityManager.getActiveNetworkInfo();
        return netInfo != null && netInfo.isConnectedOrConnecting();
    }

    private void processAuthenticationObserver(Single<Boolean> single) {
        // temporary workaround because RxLifecycle doesn't support Single
        single.toObservable()
                .compose(bindUntilEvent(ActivityEvent.DESTROY))
                .subscribe(result -> {
                    if (result) {
                        changeActivity();
                    } else {
                        if (needAnimate) {
                            animate();
                        } else {
                            Toast.makeText(this, R.string.incorrect_login_or_password, Toast.LENGTH_SHORT).show();
                        }
                    }
                });
    }

    private void changeActivity() {
        startActivity(new Intent(this, MainActivity.class));
        finish();
    }
}

File: app/src/main/kotlin/me/myshows/android/ui/login/LoginActivity.kt
package me.myshows.android.ui.login

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.text.SpannableString
import android.text.Spanned
import android.text.TextPaint
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import com.hannesdorfmann.mosby3.mvi.MviActivity
import com.jakewharton.rxbinding2.view.clicks
import io.reactivex.Observable
import kotlinx.android.synthetic.main.activity_login.*
import me.myshows.android.MyShowsApplication
import me.myshows.android.R
import me.myshows.android.ui.activities.MainActivity

class LoginActivity : MviActivity<LoginView, LoginPresenter>(), LoginView {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                    View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or
                    View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
        }

        setupNewAccountTextView(new_account)
        setEnabled(login_layout, false)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            login.setAutofillHints(View.AUTOFILL_HINT_USERNAME)
            password.setAutofillHints(View.AUTOFILL_HINT_PASSWORD)
        }
    }

    override fun loginIntent(): Observable<LoginView.Credentials> = sign_in_button.clicks()
            .map { LoginView.Credentials(login.text.toString(), password.text.toString()) }

    override fun render(state: LoginViewState) {
        when (state) {
            is AutoLoginResult -> {
                if (state.success) {
                    openMainScreen()
                } else {
                    animate()
                }
            }
            NeedManualLogin -> {
                logo.translationY = 0f
                login_layout.alpha = 1f
                setEnabled(login_layout, true)
            }
            is LoginResult -> {
                if (state.success) {
                    openMainScreen()
                } else {
                    Toast.makeText(this, R.string.incorrect_login_or_password, Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    override fun createPresenter(): LoginPresenter =
            MyShowsApplication.getComponent(this).newLoginPresenter()

    private fun setupNewAccountTextView(view: TextView) {
        val register = getString(R.string.register)
        val newAccount = getString(R.string.new_account, register)

        val start = newAccount.indexOf(register)
        val end = start + register.length

        val clickableSpan = object : ClickableSpan() {
            override fun onClick(widget: View) {
                val intent = Intent(Intent.ACTION_VIEW, Uri.parse(REGISTER_URL))
                startActivity(intent)
            }

            override fun updateDrawState(ds: TextPaint) {
                ds.color = ContextCompat.getColor(this@LoginActivity, R.color.red_80_opacity)
                ds.isUnderlineText = true
            }
        }

        val ss = SpannableString(newAccount)
        ss.setSpan(clickableSpan, start, end, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        view.setText(ss, TextView.BufferType.SPANNABLE)
        view.movementMethod = LinkMovementMethod.getInstance()
    }

    private fun openMainScreen() {
        startActivity(Intent(this, MainActivity::class.java))
        finish()
    }

    private fun setEnabled(parent: ViewGroup, enabled: Boolean) {
        parent.isEnabled = enabled
        for (i in 0 until parent.childCount) {
            val child = parent.getChildAt(i)
            when (child) {
                is ViewGroup -> setEnabled(child, enabled)
                else -> child.isEnabled = enabled
            }
        }
    }

    private fun animate() {
        val logoAnimator = ObjectAnimator.ofFloat(logo, View.TRANSLATION_Y, 0f)
        val loginLayoutAnimator = ObjectAnimator.ofFloat(login_layout, View.ALPHA, 1f)

        AnimatorSet().apply {
            playTogether(logoAnimator, loginLayoutAnimator)
            duration = ANIMATION_DURATION
            addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) = setEnabled(login_layout, true)
            })
        }.start()
    }

    companion object {
        private const val REGISTER_URL = "http://myshows.me/"
        private const val ANIMATION_DURATION: Long = 500
    }
}
